android-sdk/README.md

```diff
@@ -0,0 +1,93 @@
+
+## Problem Description
+Mobile core will define services and bindings that represent various aspects of a mobile application.  Examples of these aspects are Metrics, Authentication, and Data Synchronisation and provided by implementations such as Prometheus & Grafana, Keycloak, and fh-sync-server.  The Android core SDK will be responsible for interfacing with mobile core and providing cross cutting functionality (network request building and execution, logging, error handling, etc) while service modules will be responsible for exposing aspect implementaiton specific features and behaviors to the application module as well as consuming deployment details provided by the core SDK.
+
+During the proof of concept phase the Android mcp library and build plugin was written.  The build plugin was responsible for configuring the mcp library to connect to services that were exposed by mcp-standalone.  The application would then configure libraries for services using values that the mcp-library downloaded from mcp-standalone.  The mobile-core SDKs will follow this pattern and expand on it by also automating much of the configuration of service libraries as well as providing service libraries as first class citizens.  In the mcp PoC the developer was responsible for integrating various libraries which may have had different design patterns and goals.
+
+For example, if a KeyCloak and fh-sync-server service were created and bound, the developer was responsible for retrieving tokens from keycloak and configuring fh-sync-server to consume those tokens.  However, fh-sync-server was unaware of tokens beyond the fact that they were to be included in headers of http requests.  If a token expired fh-sync-server would return to the user a synchronization error and the developer would have had to code a system for parsing that error, understanding that it was a token timeout, refreshing the token, and restarting the sync operation.  In the mobile core world the Core SDK would understand that the sync service module connection error was a token error and invoke the keycloak module to resolve it automatically.
+
+## Proposed solution
+
+### Core SDK
+
+The Core SDK will be the package that all service modules inherit from.  It will be responsible for managing network requests, logging, error handling and resolution, loading configuration for mobile-core services, and other cross cutting concerns.  Modules will interface with the core SDK by implementing the necessary interfaces provided and annotating the implementations so that the Core SDK will be able to correctly manage them as well.  The Core SDK will also allow some of its internals to be exposed through interfaces for purposes such as testing, bespoke integrations, and management of additional resources.
+
+The Core SDK will provide a request lifecycle that the Psuedocode section will demonstrate.
+
+### Service Modules
```

Comment:
It's good to explain what module means. I understand it as separate maven dependency that you add in gradle config. Is that right?

Reply:
Yes.  I'll make a note to add a example of how it would look in code.